orttraining/orttraining/core/graph/gradient_builder_base.cc

```diff
@@ -85,17 +85,44 @@ void ComputeBroadcastBackwardAxes(
   }
 }
 
-std::vector<Dimension> GetShape(const ArgDef& arg_def) {
-  ORT_ENFORCE(arg_def.type_proto
-              && arg_def.type_proto->has_tensor_type()
-              && arg_def.type_proto->tensor_type().has_shape(),
-              "During GetShape, ", arg_def.name, "'s shape is null.");
-  std::vector<Dimension> shape;
+Status GetShape(const ArgDef& arg_def, std::vector<Dimension>& shape) {
+  //Throw OnnxruntimeShapeException type, so that it can specifically be caught in the gradient builder
+  // to build shape-independent gradient graph
+  ORT_RETURN_IF_NOT(arg_def.type_proto && arg_def.type_proto->has_tensor_type() && arg_def.type_proto->tensor_type().has_shape(),
+                    "During GetShape, ", arg_def.name, "'s shape is null.");
   const auto& dims = arg_def.type_proto->tensor_type().shape().dim();
   for (auto dim = dims.begin(); dim < dims.end(); dim++) {
     shape.push_back(*dim);
   }
-  return shape;
+  return Status::OK();
+}
+
+void ComputeBroadcastBackwardAxesDynamic(const ArgDef& a,
+                                         const ArgDef& b,
+                                         const ArgDef* a_axes,
+                                         const ArgDef* b_axes,
+                                         std::vector<NodeDef>& output) {
+  ArgDef a_shape = ArgDef("Shape_" + a.name);
```

Comment:
>ArgDef [](start = 19, length = 6)

There is still benefit of using IA(...), it will have the prefix for the forward node... making it easier to debug...